server/s3_put_multiobjectcopy_action.h

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2020 Seagate Technology LLC and/or its Affiliates
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+#pragma once
+
+#ifndef __S3_SERVER_S3_PUT_MULTIOBJECTCOPY_ACTION_H__
+#define __S3_SERVER_S3_PUT_MULTIOBJECTCOPY_ACTION_H__
+
+#include <memory>
+
+#include "s3_object_action_base.h"
+#include "s3_async_buffer.h"
+#include "s3_bucket_metadata.h"
+#include "s3_motr_writer.h"
+#include "s3_object_metadata.h"
+#include "s3_part_metadata.h"
+#include "s3_probable_delete_record.h"
+#include "s3_object_data_copier.h"
+
+enum class S3PutPartCopyActionState {
+  empty = 0,         // Initial state
+  validationFailed,  // Any validations failed for request, including metadata
+  probableEntryRecordFailed,
+  newObjOidCreated,         // New object created
+  newObjOidCreationFailed,  // New object create failed
+  writeComplete,            // data write to object completed successfully
+  writeFailed,              // data write to object failed
+  md5ValidationFailed,      // md5 check failed
+  metadataSaved,            // metadata saved for new object
+  metadataSaveFailed,       // metadata save failed for new object
+  completed,                // All stages done completely
+};
+
+class S3PutMultiObjectCopyAction : public S3ObjectAction {
```

Comment:
This class should inherit S3PutObjectActionBase as per the design as suggestion from Rajesh. Also, we can omit put from class name.